src/main/java/net/snowflake/ingest/streaming/internal/ParquetFlusher.java

```diff
@@ -0,0 +1,278 @@
+    List<List<Object>> rows = null;
+
+    for (ChannelData<ParquetRowBuffer.ParquetChunkData> data : channelsDataPerTable) {
+      // Create channel metadata
+      ChannelMetadata channelMetadata =
+          ChannelMetadata.builder()
+              .setOwningChannel(data.getChannel())
+              .setRowSequencer(data.getRowSequencer())
+              .setOffsetToken(data.getOffsetToken())
+              .build();
+      // Add channel metadata to the metadata list
+      channelsMetadataList.add(channelMetadata);
+
+      logger.logDebug(
+          "Parquet Flusher: Start building channel={}, rowCount={}, bufferSize={} in blob={}",
+          data.getChannel().getFullyQualifiedName(),
+          data.getRowCount(),
+          data.getBufferSize(),
+          filePath);
+
+      if (rows == null) {
+        columnEpStatsMapCombined = data.getColumnEps();
+        rows = new ArrayList<>();
+        firstChannel = data.getChannel();
+      } else {
+        // This method assumes that channelsDataPerTable is grouped by table. We double check
+        // here and throw an error if the assumption is violated
+        if (!data.getChannel()
+            .getFullyQualifiedTableName()
+            .equals(firstChannel.getFullyQualifiedTableName())) {
+          throw new SFException(ErrorCode.INVALID_DATA_IN_CHUNK);
+        }
+
+        columnEpStatsMapCombined =
+            ChannelData.getCombinedColumnStatsMap(columnEpStatsMapCombined, data.getColumnEps());
+      }
+      rows.addAll(data.getVectors().rows);
+
+      rowCount += data.getRowCount();
+
+      logger.logDebug(
+          "Parquet Flusher: Finish building channel={}, rowCount={}, bufferSize={} in blob={}",
+          data.getChannel().getFullyQualifiedName(),
+          data.getRowCount(),
+          data.getBufferSize(),
+          filePath);
+    }
+
+    Map<String, String> metadata = channelsDataPerTable.get(0).getVectors().metadata;
+
+    flushToParquet(chunkData, rows, metadata);
+    return new SerializationResult(channelsMetadataList, columnEpStatsMapCombined, rowCount);
+  }
+
+  private void flushToParquet(
+      ByteArrayOutputStream output, List<List<Object>> rows, Map<String, String> metadata)
+      throws IOException {
+    try {
+      ParquetWriter<List<Object>> writer =
+          new BdecParquetWriterBuilder(output, schema, metadata)
+              .withWriterVersion(ParquetProperties.WriterVersion.PARQUET_1_0)
+              .withDictionaryEncoding(false)
+              .enableValidation()
+              .withCompressionCodec(CompressionCodecName.UNCOMPRESSED)
+              .withWriteMode(ParquetFileWriter.Mode.CREATE)
+              .build();
+
+      for (List<Object> row : rows) {
+        writer.write(row);
+      }
+      writer.close();
+    } catch (Throwable t) {
+      logger.logError("Parquet Flusher: failed to write", t);
+      throw t;
+    }
+  }
+
+  public static class BdecParquetWriterBuilder
```

Comment:
This might be better segmented in its own file with a proper docstring. Also, why is it public? Isn't it only used here?

Reply:
Indeed it can be a separate class. My point was that these classes are fairly small and used only here.
we can move them out if everything grows too much or/and we want them to be unit tested.
This one can indeed be private, good point.